sql/src/main/java/io/crate/expression/scalar/cast/CastFunctionResolver.java

```diff
@@ -21,70 +21,80 @@
 
 package io.crate.expression.scalar.cast;
 
-import io.crate.common.collections.Lists2;
 import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Literal;
 import io.crate.expression.symbol.Symbol;
 import io.crate.metadata.FunctionIdent;
 import io.crate.metadata.FunctionInfo;
 import io.crate.types.ArrayType;
 import io.crate.types.DataType;
-import io.crate.types.ObjectType;
 
 import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 
-import static io.crate.types.DataTypes.GEO_POINT;
-import static io.crate.types.DataTypes.GEO_SHAPE;
-import static io.crate.types.DataTypes.PRIMITIVE_TYPES;
+import static io.crate.types.DataTypes.OBJECT;
 
 public class CastFunctionResolver {
 
     static final String TRY_CAST_PREFIX = "try_";
     private static final String TO_PREFIX = "to_";
 
-    static final Map<DataType, String> CAST_SIGNATURES; // data type -> name
-
-    static {
-        List<DataType> CAST_FUNC_TYPES = Lists2.concat(
-            PRIMITIVE_TYPES,
-            List.of(GEO_SHAPE, GEO_POINT, ObjectType.untyped()));
-
-        CAST_SIGNATURES = new HashMap<>((CAST_FUNC_TYPES.size()) * 2);
-        for (var type : CAST_FUNC_TYPES) {
-            CAST_SIGNATURES.put(type, castFuncName(type));
-
-            var arrayType = new ArrayType<>(type);
-            CAST_SIGNATURES.put(arrayType, castFuncName(arrayType));
-        }
-    }
-
-    private static String castFuncName(DataType type) {
-        return TO_PREFIX + type.getName();
-    }
-
     public static Symbol generateCastFunction(Symbol sourceSymbol, DataType targetType, boolean tryCast) {
         DataType sourceType = sourceSymbol.valueType();
-        FunctionInfo functionInfo = functionInfo(sourceType, targetType, tryCast);
-        return new Function(functionInfo, List.of(sourceSymbol));
+        if (ArrayType.unnest(targetType).id() == OBJECT.id()) {
+            // Currently, it is not possible to resolve a function based on
+            // its return type which is required to create the `to_object` cast
+            // function that would be able to utilize the inner object types.
+            // Therefore, we register a dedicated `to object` cast function with
+            // two arguments, where the second argument carries information about
+            // the object target type and its inner types.
+            return new Function(
+                functionInfo(List.of(sourceType, targetType), targetType, tryCast),
+                // the null literal is passed as an argument to match the method signature
+                List.of(sourceSymbol, Literal.NULL));
+        } else {
+            return new Function(
+                functionInfo(List.of(sourceType), targetType, tryCast),
+                List.of(sourceSymbol));
+        }
```

Comment:
Would it make sense to always encode the return type into the cast function and generally remove the notion of having `per type` cast function names?

So that we don't have to treat object casts in a special way.

Reply:
we can only provide the return type as a function argument, because at this step https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/expression/BaseImplementationSymbolVisitor.java#L56 when we resolve a function the return type info from the `FunctionInfo` is discarded. Thus, we would then need to make all the cast functions to take two arguments (source symbol and target type), and build `Function` with arguments like `List.of(sourceSymbol, Literal.NULL))`. But actually then we don't have to distinguish between cast and cast object functions. So do you suggest to go with the changing of the cast function signatures to take two args?

Reply:
>  So do you suggest to go with the changing of the cast function signatures to take two args?

Yes. It is also a hack, but so is encoding the return type into the function name. And I think we're better off with a single hack than multiple hacks + using a different hack depending on which type.

Reply:
But we'll probably have to make sure that if a <4.0.12 node creates a plan with a old cast function name, it will still resolve on a node that includes this fix.

Reply:
maybe smth like that 4d7dcc32ff86e7668a2821a5250379546d45e288. it also stay backward compatible, coz resolvers either pick the provided pre-registered return data type or get it from the data types.